# PANDAS LAB - Continuation of the EuroLeague Case Study
## Let's compare the competitions!

![](https://eurospects.com/wp-content/uploads/2018/10/eurocupeuroleague.png)

First of all, we must 

- Import the actual.csv Dataset. This dataset will be the CSV of the result obtained in class.
- Import ec23.xlsx. This dataset correspond to the import the dataset corresponding to Eurocup. Don't worry, this time the Excel will be cleaned (almost)

<details>
    <summary> Lab results </summary>

![](https://media1.tenor.com/m/lx2WSGRk8bcAAAAC/pulp-fiction-john-travolta.gif)

PERO QUÉ OS HABÉIS CREÍDO

</details>

In [ ]:
#pip install openpyxl
# In case there is an error with this .xlsx document opening, install this library

In [1]:
import pandas as pd
import numpy as np

/var/folders/2v/nsbsx6y90_9_6kks00t6shb80000gn/T/ipykernel_27206/2162656668.py:1: DeprecationWarning: 
Pyarrow will become a required dependency of pandas in the next major release of pandas (pandas 3.0),
(to allow more performant data types, such as the Arrow string type, and better interoperability with other libraries)
but was not found to be installed on your system.
If this would cause problems for you,
please provide us feedback at https://github.com/pandas-dev/pandas/issues/54466
        
  import pandas as pd


In [39]:
elteams=pd.read_csv("datasets/actual.csv")
ecteams=pd.read_excel("datasets/ec2023.xlsx", index_col=False)

Let's analyze the EuroCup Dataframe! Is there any cleaning needed? Remember the basics: 
- Drop Null Values
- Drop Duplicate Rows
- Drop Strange Columns if you see any

In [40]:
# Your Code here: 

ecteams.shape


(173, 10)

In [41]:
ecteams.head()

Unnamed: 0                fecha  ronda                        local  \
0           0  2024-01-31T21:00:00     17       Dreamland Gran Canaria   
1           1  2024-01-31T20:00:00     17      Dolomiti Energia Trento   
2           2  2024-01-31T19:30:00     17               ratiopharm Ulm   
3           3  2024-01-31T18:30:00     17  Cedevita Olimpija Ljubljana   
4           4  2024-01-31T18:00:00     17          Turk Telekom Ankara   

                        visitante  localscore  roadscore     Winner  Ganador  \
0                U-BT Cluj-Napoca         111      113.0  Visitante    113.0   
1        Buducnost VOLI Podgorica          78       97.0  Visitante     97.0   
2         Aris Midea Thessaloniki          66       86.0  Visitante     86.0   
3                    London Lions          85       92.0  Visitante     92.0   
4  Mincidelice JL Bourg en Bresse          64       68.0  Visitante     68.0   

   plusminus  
0       -2.0  
1      -19.0  
2      -20.0  
3       -7.0  
4       -4.0

In [42]:
ecteams.drop(columns='Unnamed: 0', inplace = True) #Lets remove the not needed columns
ecteams.sample(2)

fecha  ronda                  local       visitante  \
28   2024-01-17T18:00:00     15  Prometey Slobozhanske  Wolves Vilnius   
100  2023-11-21T19:30:00      8  Veolia Towers Hamburg  Wolves Vilnius   

     localscore  roadscore     Winner  Ganador  plusminus  
28           91       64.0      Local     91.0       27.0  
100          71       84.0  Visitante     84.0      -13.0

In [43]:
ecteams.isnull().sum() # Check if there are any null values

fecha         0
ronda         0
local         0
visitante     0
localscore    0
roadscore     1
Winner        0
Ganador       1
plusminus     1
dtype: int64

In [44]:
ecteams[ecteams.isnull().any(axis=1)] #Lets see which game has null values

fecha  ronda                  local  \
142  2023-10-24T19:30:00      4  Veolia Towers Hamburg   

                  visitante  localscore  roadscore     Winner  Ganador  \
142  Hapoel Shlomo Tel Aviv          72        NaN  Visitante      NaN   

     plusminus  
142        NaN

In [48]:
ecteams.dropna(inplace=True) #We remove the row

In [49]:
ecteams.isnull().sum() # Check to verify we have a clean data of nulls

fecha         0
ronda         0
local         0
visitante     0
localscore    0
roadscore     0
Winner        0
Ganador       0
plusminus     0
dtype: int64

In [52]:
ecteams.duplicated().sum() # We have two duplicated

2

In [51]:
ecteams[ecteams.duplicated()]

fecha  ronda                           local  \
80  2023-12-05T19:30:00     10           Veolia Towers Hamburg   
99  2023-11-21T19:30:00      8  Mincidelice JL Bourg en Bresse   

                visitante  localscore  roadscore     Winner  Ganador  \
80  Prometey Slobozhanske          83       97.0  Visitante     97.0   
99    Turk Telekom Ankara          86       67.0      Local     86.0   

    plusminus  
80      -14.0  
99       19.0

In [55]:
ecteams.drop_duplicates(inplace=True) # Remove and check
ecteams.duplicated().sum()

0

Let's now compare it with the EL dataset that we have: You might select a sample of both of them and see what are the inconsistencies!

In [56]:
#Your sample ecteams code here: 
ecteams.sample (5)

fecha  ronda                       local  \
102  2023-11-21T18:00:00      8            U-BT Cluj-Napoca   
29   2024-01-16T20:00:00     15     Dolomiti Energia Trento   
159  2023-10-10T20:00:00      2     Dolomiti Energia Trento   
20   2024-01-17T21:00:00     15      Dreamland Gran Canaria   
7    2024-01-30T18:00:00     17  7bet-Lietkabelis Panevezys   

                          visitante  localscore  roadscore     Winner  \
102          Dreamland Gran Canaria          70       92.0  Visitante   
29                    Slask Wroclaw          85       79.0      Local   
159  Mincidelice JL Bourg en Bresse          72       79.0  Visitante   
20              Turk Telekom Ankara          73       83.0  Visitante   
7                     Slask Wroclaw          91       75.0      Local   

     Ganador  plusminus  
102     92.0      -22.0  
29      85.0        6.0  
159     79.0       -7.0  
20      83.0      -10.0  
7       91.0       16.0

In [58]:
ecteams.shape

(170, 9)

In [57]:
#Your sample elteams code here: 
elteams.sample(5)

Competition           Phase  Ronda         Dia      Hora  \
63   Euroleague  Regular Season      8  2023-11-14  19:00:00   
189  Euroleague  Regular Season     22  2024-01-18  18:30:00   
223  Euroleague  Regular Season     25  2024-02-02  20:30:00   
203  Euroleague  Regular Season     23  2024-01-26  18:30:00   
66   Euroleague  Regular Season      8  2023-11-14  20:30:00   

                        Local                      Visitante  Localscore  \
63                  AS Monaco          Anadolu Efes Istanbul          82   
189     Anadolu Efes Istanbul                   FC Barcelona          98   
223  Virtus Segafredo Bologna  Partizan Mozzart Bet Belgrade          88   
203     Anadolu Efes Istanbul                      AS Monaco          78   
66            Valencia Basket                    Real Madrid          73   

     Roadscore                                           Arbitros  Plusminus  \
63          89  ['PUKL,SASA', 'KARDUM, LUKA', 'UDYANSKYY, EDUA...         -7   
189         74  ['HORDOV, TOMISLAV ', 'FOUFIS, IOANNIS', 'TSAR...         24   
223         84  ['LOTTERMOSER, ROBERT', 'PUKL,SASA', 'TRAWICKI...          4   
203         80  ['GARCIA, JUAN CARLOS', 'ROCHA, FERNANDO', 'VA...         -2   
66          76  ['RADOVIC, SRETEN', 'LAURINAVICIUS, JURGIS', '...         -3   

                       Winner    Ganador  
63      Anadolu Efes Istanbul  Visitante  
189     Anadolu Efes Istanbul      Local  
223  Virtus Segafredo Bologna      Local  
203                 AS Monaco  Visitante  
66                Real Madrid  Visitante

In [59]:
elteams.shape

(225, 13)

What are the inconsistencies that you are seeing? Write them down!

- Table sizes are different (9 columns vs 13 columns)
 - Fecha is divided in two (elteams) (different data type)
 - Competition, Arbitros and Phase are not in (enteams)
- Column names are different (upper case vs lower case)
- Columns have a different order
- elteams have 225 rows vs 170 rows in enteams
- Winner is showing different things


<details>
    <summary> Expand the box to see the inconsistencies </summary>

- The columns are named in lowercase and the columns in Euroleague are capitalized
- There are some missing columns (f.e. Arbitros)
- Fecha is not split in Dia and Hora
- `Winner and Ganador Columns show different data` are not the same! Be extremely careful about it

</details>

As you can see there are a lot of different inconsistencies right? 
But beofore starting cleaning like crazy people, let's take a second to know what do we want to compare: 

- The CLUBS are impossible to compare, as different competitions defines different clubs
- Both rounds (Eurocup has less rounds) and Dates (Different Dates) are also not really useful
- We don't have information about Arbitros

## Conclusion

### We must take into consideration what data we DO have in both dataframes to compare it propperly and don't use time cleaning unnecessary things. 

In this particular case, we will need to compare: 


`Competition`: The main column that define one from another. 

Create the Competition column in the EuroCup dataframe, being the fixed value "Eurocup"

In [61]:
#Your Code here: 
ecteams ['Competition'] = 'Eurocup'
ecteams.sample(2)

fecha  ronda              local              visitante  \
130  2023-10-31T20:30:00      5  Joventut Badalona  Veolia Towers Hamburg   
22   2024-01-17T19:30:00     15     ratiopharm Ulm       U-BT Cluj-Napoca   

     localscore  roadscore     Winner  Ganador  plusminus Competition  
130         108       76.0      Local    108.0       32.0     Eurocup  
22           81       85.0  Visitante     85.0       -4.0     Eurocup

`Ganador`: Being the actual values "Local" or "Visitante"

**BE CAREFUL**: Before changing the correct column to "Ganador" from the EC dataframe, there is a "Ganador" column with the incorrect information: STEP BY STEP!

In [66]:
#Drop the Column Ganador from the Eurocup Dataframe
ecteams.drop(columns= 'Ganador', inplace=True)
ecteams.sample(2)


fecha  ronda                           local  \
163  2023-10-05T21:00:00      1  Mincidelice JL Bourg en Bresse   
3    2024-01-31T18:30:00     17     Cedevita Olimpija Ljubljana   

         visitante  localscore  roadscore     Winner  plusminus Competition  
163  Slask Wroclaw          90       74.0      Local       16.0     Eurocup  
3     London Lions          85       92.0  Visitante       -7.0     Eurocup

In [67]:
#Rename the column with the actual information to "Ganador"
ecteams.rename(columns = {'Winner':'Ganador'}, inplace=True)
ecteams.sample(2)

fecha  ronda                       local  \
58   2023-12-19T18:00:00     12  7bet-Lietkabelis Panevezys   
134  2023-10-25T20:00:00      4               Slask Wroclaw   

                   visitante  localscore  roadscore    Ganador  plusminus  \
58            ratiopharm Ulm          94       90.0      Local        4.0   
134  Aris Midea Thessaloniki          63       80.0  Visitante      -17.0   

    Competition  
58      Eurocup  
134     Eurocup

As seen in the first look at the dataframes, there was a problem with the columns: Are lower when they should be Capitalized, as in the Euroleague DataFrame. Change the column names to capitalize. 

is there a quick way to do it?

In [87]:
ecteams.columns = map(lambda x: x.capitalize(), ecteams.columns)
ecteams.columns

Index(['Fecha', 'Ronda', 'Local', 'Visitante', 'Localscore', 'Roadscore',
       'Ganador', 'Plusminus', 'Competition'],
      dtype='object')

Now it is time to just take the useful columns from each data! We will call the result subsets `comp_el` and `comp_ec`

REMEMBER: Use the same order of the columns for each dataset. 

Recommended order: Competition, Localscore, Roadscore, Plusminus, Ganador. **Nothing else is useful**

In [94]:
# Your code here: 
comp_ec = ecteams.loc[:,['Competition','Localscore','Roadscore', 'Plusminus', 'Ganador']]
comp_ec.sample(2)


Competition  Localscore  Roadscore  Plusminus Ganador
7      Eurocup          91       75.0       16.0   Local
84     Eurocup          69       67.0        2.0   Local

In [96]:
comp_el = elteams.loc[:, ['Competition','Localscore','Roadscore', 'Plusminus', 'Ganador']]
comp_el.sample(2)

Competition  Localscore  Roadscore  Plusminus    Ganador
145  Euroleague          84         89         -5  Visitante
104  Euroleague          91         94         -3  Visitante

Time to concatenate them both! Create a concat of both Dfs, called `competition`

In [99]:
# Your code here: 
competition = pd.concat ([comp_el, comp_ec])
competition.sample(4)

Competition  Localscore  Roadscore  Plusminus    Ganador
105  Euroleague          80       75.0        5.0      Local
118     Eurocup          69       82.0      -13.0  Visitante
154  Euroleague          77       71.0        6.0      Local
27      Eurocup          93      104.0      -11.0  Visitante

## Well done!

Now we have the combination of both competitions in the same dataframe. Is time for the questions you must answer: 

1. How many matches have been in each competition?
2. What are the maximum points that a local team has scored per competition? And the road team as well. 
3. Do the same operations with the average points (per competition)
4. Can you tell me the maximum difference that has been in a match, splitted by competition? 
    - Careful: The Plusminus is both positive and negative, being negative that the Road Team has achieved the Victory. To have the absolute, create a new Column named Absolut_Plusminus. Remember: np.where is life!
5. Also grouped by competition, give me the number of victories of the Locals and the Visitors

In [103]:
# 1 Number of matches for each competition
competition.groupby('Competition').count() # We asume that the total teams that have Win is the total number of games of each competition

Localscore  Roadscore  Plusminus  Ganador
Competition                                           
Eurocup             170        170        170      170
Euroleague          225        225        225      225

In [105]:
# 2 Max score for Local and Road for each competition
competition.groupby('Competition').agg({'Localscore':'max', 'Roadscore':'max'})

Localscore  Roadscore
Competition                       
Eurocup             120      114.0
Euroleague          130      126.0

In [106]:
# 3 Average score for each competition
competition.groupby('Competition').agg({'Localscore':'mean', 'Roadscore':'mean'})

Localscore  Roadscore
Competition                       
Eurocup       85.235294  83.958824
Euroleague    83.284444  79.173333

In [115]:
# 4 Maximum score difference in games
competition['Absolut_Plusminus'] = np.where(competition['Plusminus']<0,competition['Plusminus'].abs(),competition['Plusminus'])
competition.sort_values(by = ['Absolut_Plusminus'], ascending=False )

Competition  Localscore  Roadscore  Plusminus    Ganador  \
35   Euroleague          98       59.0       39.0      Local   
141     Eurocup          72      111.0      -39.0  Visitante   
104     Eurocup          60       98.0      -38.0  Visitante   
146     Eurocup          69      105.0      -36.0  Visitante   
162  Euroleague         109       74.0       35.0      Local   
..          ...         ...        ...        ...        ...   
147  Euroleague          76       77.0       -1.0  Visitante   
149  Euroleague          85       84.0        1.0      Local   
76   Euroleague          74       73.0        1.0      Local   
138     Eurocup          92       93.0       -1.0  Visitante   
90   Euroleague         100       99.0        1.0      Local   

     Absolut_Plusminus  
35                39.0  
141               39.0  
104               38.0  
146               36.0  
162               35.0  
..                 ...  
147                1.0  
149                1.0  
76                 1.0  
138                1.0  
90                 1.0  

[395 rows x 6 columns]

In [113]:
competition.groupby('Competition').agg({'Absolut_Plusminus':'max'})

Absolut_Plusminus
Competition                   
Eurocup                   39.0
Euroleague                39.0

In [118]:
# 5 number of victories of the Locals and the Visitors
pd.crosstab(competition['Competition'], competition ['Ganador'])

Ganador      Local  Visitante
Competition                  
Eurocup         90         80
Euroleague     150         75